src/common/Throttle.h

```diff
@@ -330,4 +331,183 @@ class OrderedThrottle {
+      }
+
+      assert(c >= 0);
+      if (c) {
+	// put c tokens into bucket
+	int64_t current = remain;
+	if ((current + c) <= (int64_t)max) {
+	  remain += c;
+	} else {
+	  remain = (int64_t)max;
+	}
+      }
+      return remain;
+    }
+
+    void set_max(int64_t m) {
+      if (remain > m)
+	remain = m;
+      max = m;
+    }
+  };
+
+  struct Blocker {
+    int64_t tokens_requested;
+    T *got_token_handler;
+    I *item;
+
+    Blocker(int64_t _tokens_requested, T *handler, I *_item)
+    : tokens_requested(_tokens_requested), got_token_handler(handler),
+      item(_item) {}
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  int64_t m_avg = 0;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+  list<Blocker*> m_blockers;
+  Mutex m_lock;
+
+public:
+TokenBucketThrottle(
+    CephContext *cct,
+    int64_t capacity,
+    int64_t avg,
+    SafeTimer *timer,
+    Mutex *timer_lock)
+  : m_cct(cct), m_throttle(m_cct, "token_bucket_throttle", capacity),
+    m_avg(avg), m_timer(timer), m_timer_lock(timer_lock),
+    m_lock("token_bucket_throttle_lock")
+{
+  Mutex::Locker timer_locker(*m_timer_lock);
+  schedule_timer();
+}
+
+~TokenBucketThrottle()
+{
+  {
+    Mutex::Locker timer_locker(*m_timer_lock);
+    cancel_timer();
+  }
+  while (!m_blockers.empty()) {
+    Blocker *b = m_blockers.front();
+    m_blockers.pop_front();
+    m_lock.Unlock();
+    b->got_token_handler->handle(0, b->item);
+    m_lock.Lock();
+    delete b;
+  }
+}
+
+bool get(int64_t c, T *handler, I *item) {
+  if (0 == m_throttle.max)
+    return false;
+
+  assert(c >= 0);
+  bool waited = false;
+
+  Mutex::Locker lock(m_lock);
+  int64_t got = m_throttle.get(c);
+  if (got < c) {
+    // Got not enough tokens, add a blocker for it.
+    m_blockers.push_back(new Blocker(c - got, handler, item));
+    waited = true;
+  }
+  return waited;
+}
+
+void set_max(int64_t m) {
+  Mutex::Locker lock(m_lock);
+  m_throttle.set_max(m);
+}
+
+void set_average(int64_t avg) {
+  m_avg = avg;
+}
+
+private:
+void add_tokens() {
+  // check the m_blockers from head to tail, if blocker can get
+  // enough tokens, let it go.
+  while (!m_blockers.empty()) {
+    Blocker *blocker = m_blockers.front();
+    int64_t got = m_throttle.get(blocker->tokens_requested);
+    if (got == blocker->tokens_requested) {
+      // got enough tokens for front.
+      m_blockers.pop_front();
+      m_lock.Unlock();
+      m_timer_lock->Unlock();
+      blocker->got_token_handler->handle(0, blocker->item);
+      m_timer_lock->Lock();
+      m_lock.Lock();
+      delete blocker;
+    } else {
+      // there is no more tokens.
+      blocker->tokens_requested -= got;
+      break;
+    }
+  }
+}
+
+void schedule_timer() {
+  {
+    Mutex::Locker lock(m_lock);
+    m_throttle.put(m_avg);
+    add_tokens();
+  }
+  m_token_ctx = new FunctionContext(
+      [this](int r) {
+        schedule_timer();
+      });
+
+  m_timer->add_event_after(1, m_token_ctx);
```

Comment:
This is a very course timer -- if I inject 200 IOPs in a 100 IOPS throttled bucket, it will result in very bursty behavior instead of releasing the next IO when the next tocken (smoothly) becomes available. Taking the leaking bucket example (again, this implementation seems to be backwards since the bucket to leak instead of be filled), the leak isn't going to "drip" a large quantity of water, stop leaking for 1 second, and repeat. 

The nice thing about reversing your model is that you would only need to schedule a timer if the bucket becomes full.

Reply:
Do you mean leakybucket? I am not sure I understand this comment correctly. But I don't think we can only schedule a timer if the bucket becomes full. Because we don't know whether it is full after ```ONE second```, when this event will be actually working.

Reply:
Yes -- using the concept of a leaky bucket where "served" IOPS are leaking at a fixed rate and "incoming" IOPS are filling the bucket, the leak rate is smoothed. From what I can figure with this current implementation, injecting higher IOPS can the maximum will resulting 1 sec periodic bursting behavior (i.e. release all IOPS allowed immediately) to the backing OSDs instead of a smoothed "leak".

Reply:
About the leaky bucket vs token bucket, the reason I chose token bucket because I don't need to control the actual request at all but only control the token in bucket. 

But as I explained in my one comment, I am okey some one to implement a leaky bucket if needed, that's why I did not put Class Bucket into Class TockenBucket, because there would be a class of LeakyBucket will use it. 

In my opinion, both of leakybucket and tokenbucket can do a iops limiting here.